### ISSUE TITLE
Allow buildings to assist construction

### ISSUE THREAD

**Issue by WatchTheFort** [beyond-all-reason/spring#1597]
Currently a building with buildpower cannot assist other construction, only perform its own i.e. be a factory.

Due to this, construction turrets are implemented as mobile units with zero speed, but this has unwanted side effects, e.g. radar ghosts do not appear in fog because they are technically not buildings.

Allow buildings with buildpower to assist other construction, supporting the full set of building unit def tags that mobile builders use.

One consequence of this is that we will need to decide what counts as factory. One option is a factory is a unit with buildpower, a buildlist, and can start construction within its own footprint. Thus, SupCom's Fatboy would count as a factory, since it builds units inside itself, but BAR's combat engineers wouldn't, since they build units beside themselves.

**Comment by sprunk** [comment]
Sounds like what you really want is to have the unwanted side effects fixed.

**Comment by WatchTheFort** [comment]
> Sounds like what you really want is to have the unwanted side effects fixed.

Since the implementation itself is a workaround, it would be nicer to not have to implement additional workarounds for the workaround.

**Comment by saurtron** [comment]
Took a look at this, and one problem I see is in how orders are given, right now factories have most/all of the non unit creation commands assigned to the built unit queue.

Having factories support the same commands would make it a bit problematic to know where the queued command needs to be put. For example, if a guard/repair command is given to the factory, should it be given to the to be built units, or to the factory itself. Not sure how gui could do that.

**Comment by WatchTheFort** [comment]
What about for buildings that aren't factories, i.e. construction turrets?

For UI for factories, we could split the command palette in two, one for the factory itself, one for its waypoints.

**Comment by sprunk** [comment]
Add `CMD.OPT_PUT_ORDER_INTO_THE_OPPOSITE_QUEUE` cmdoption flag which puts orders in the opposite queue compared to where they would normally end up.

**Comment by saurtron** [comment]
> What about for buildings that aren't factories, i.e. construction turrets?

For those it wouldn't be a problem, but I think there's no building ai class atm, main distinction now is movable or factory, maybe we could introduce an immobile AI to be the parent of factory, or make some specialization of factory, or just not do anything special in that case, but probably would be good... not sure if needed at this point yet.

**Comment by saurtron** [comment]
I got some advances here, but would like some feedback.

I have a working implementation of factories able of guard, repair and fight commands. You can check [this branch](https://github.com/saurtron/spring/tree/factory-builder-commands) or [this commit](https://github.com/saurtron/spring/commit/b3acf9aa06a7f2133e2c3c6e61baefd1e6225db5) to see more or less what this needs. This is just a proof of concept implementation, so please don't roast me XD.

What this shows is basically a lot of code from Builder.cpp and BuilderCAI.cpp is needed with pretty minimal changes, just removing some parts where movement is added. Adding resurrect and terraform will add even more code but shouldn't be much more difficult.

Of course all of this needs to be refactored to share code. Also, the commands should work for any buildings, not just for factories, I just merged them there to have a feeling of the difficulty.

Code now is normally shared among AIs and unit types through inheritance, and there's high coupling among unit types and cais, specially for Builder and Factory types.

The problem I see, is how the inheritance graphs (for unti types and command AIs) are constructed at the moment:

![graphs](https://github.com/user-attachments/assets/8e843f75-6203-47eb-921d-61c7bffdf0d3)

Right now, I think buildings can be using any of `FactoryCAI` or `CommandAI` for ai, and `Building` subtypes for unit type.

Regarding command AI, ideally, there would be some code that could be shared among FactoryCAI and BuilderCAI (and maybe a new type BuildingCAI, for buildings not requiring to be factories but wanting build command support like [metal extractors](https://github.com/beyond-all-reason/Beyond-All-Reason/pull/3995#issue-2718748453)).

For unit types it could be more simple since there's already a Building type that could probably be used, although a BuilderBuilding could be included between Building and Factory.

Would like to hear feedback, or ideas for other possible approaches.

--

teaser screenshot with working "building" nanoturrets (using Factory type and FactoryCAI atm).

![nanofactories](https://github.com/user-attachments/assets/a5f094fb-07f3-4505-855a-3bf3c48edb74)

Yes, they look the same :P, the only advantage atm is after getting transported they always end in cardinal direction instead of sometimes getting slightly rotated XD.

**Comment by sprunk** [comment]
Extracting `CBuilderCache` sounds like an excellent minimal starter refactor PR that could be merged right away. See also #1888. (Haven't had the time to look much beyond these, but minimal refactors like these two would be best brought individually to the front as well)

**Comment by saurtron** [comment]
Since those action implementations don't seem to be extremely coupled among themselves, also metal extractor behaviour seems to be pretty loosely coupled... I'm thinking maybe trying a composition approach could be worth it, where some sort of Behaviour/Action base class is created that can be added to any CommandAI independently.

So, these are the ways I see CommandAIs could be refactored to allow for builder commands, and possibly also extractor behaviour, to be used in more CommandAIs:

1. Restructure the **single inheritance** graph.
   - Might need to have support for too many uneeded things for unrelated units/commandAis.
     - For example allowing builder commands for movable builders and buildings can be tricky without also giving support to MobileAIs in general
     - Another example is the metal extractor + assist turret case
1. **Multiple inheritance**.
   - Can get messy quick and require too many class compositions
   - Not sure this is allowed due to desync concerns
1. **Composition** with some sort of Behaviour/CommandAIComponent class that can be pushed into CommandAI/UnitSubclass to better be able to extend.
   - Might not be feasible or messy if behaviours prove to be too tightly coupled
   - If it works could end up being a lot more extensible and clean
1. Allowing a unit to have **several CommandAIs**
   - Just adding for sake of completeness here, but don't think this would be a good approach

**Comment by burnhamrobertp** [comment]
Read through the unit type and CommandAI code for this and checked the key bits in an engine-only test game. Mobile vs static does not look like the axis to restructure around; BuilderCAI already runs fine on immobile builders. What seems to the be the issue - correct me if I'm wrong - relates back to the shape of the inheritance of object types (better represented in @saurtron 's visualization above):

```mermaid
flowchart TD
  U[CUnit] --> B[CBuilder<br/>assist, repair, reclaim]
  U --> S[CBuilding<br/>yardmap, blockMap]
  S --> X[CExtractorBuilding<br/>metal squares]
  S --> F[CFactory<br/>builds inside itself]
```

CBuilder and CBuilding are siblings, so nothing can be structure + builder or extractor + builder. Oddly - to me anyways - is that a factory is determined by the yardmap and having build power, alone? This seems both a limiting assumption and kind of unnecessary/indirect? `IsFactoryUnit()` is `IsBuilderUnit() && IsBuildingUnit()`, so a static builder with a yardmap becomes a CFactory with canAssist forced off. You can take any static builder def and flip it from CBuilder / BuilderCAI to CFactory / FactoryCAI purely on adding yardMap = "oooo".

This suggests that there may be an elegant way forward, and using the approach already patterened by moveType and weapons:

```mermaid
flowchart LR
  subgraph CUnit [CUnit, one pool slot]
    direction TB
    D[structure data<br/>yardmap, blockMap, blockHeightChanges]
    subgraph C [capabilities, from explicit def tags]
      direction LR
      M[Move]
      W[Weapons]
      B[Builder]
      E[Extractor]
      F[Factory]
    end
  end
```

Structure stays plain data on CUnit gated on `IsBuildingUnit()`. Builder, Extractor and Factory become things a unit owns, created from explicit def tags rather than inferred, with SanitizeUnitDefs rejecting combinations the engine cannot do instead of quietly picking a class.

It seems like there's a path to do the work in reasonable slices too? Similar to https://github.com/beyond-all-reason/RecoilEngine/pull/1888 & [#1890](https://github.com/beyond-all-reason/RecoilEngine/pull/1890). 

1. fold CBuilding into CUnit
2. add an explicit factory tag replacing the yardmap inference, defaulting to today's behavior
3. extractor behavior actually tied to the def, rather than the class. 1+2 gives the construction turret this issue opens with; 3 allows that building to be a mex to. FactoryCAI's two queues, the canAssist veto and where GUARD goes stay out of it; that is the factory-assist half and its own discussion.

Does this approach seem feasible? Or is there some element of the unit definition and behavior that is problematic? Refactoring the behavior to be driven not by unit type, but by the presence or absence of these defs is actually the larger change, I think, but surprisingly the thing that seemed the highest risk (precluding combinations of behaviors that aren't accounted for / aren't desired in units) is easier than I had first expected.

**Comment by sprunk** [comment]
Mexes are a subset of buildings because it is not well-defined what should happen when they move. One could say "the same thing that currently happens if a building!mex dies at the previous position, and a new one is created at the new position", but there are still large issues with that (order of events matters for overlap while units move kinda simultaneously, the operation is potentially expensive). Then there's the larger issue where native mexes (regardless if mobile or not) implement just some specific game design which doesn't give games much control (what if I don't want the area to be a circle? what if I don't want to tie it to activation? what if i want overlapping mexes to stack? to stack partially? what about extracting other resources? what about stacking if the resources are different? do different resource get different metalmap? etc). Personally my vision for mexes would be to expose a Lua interface that lets a game implement their own (as described in https://github.com/beyond-all-reason/RecoilEngine/pull/1322#issuecomment-2580893859) and move towards deprecating the native mexes and eventually getting rid of them. None of this precludes making mexes into a capability until then though, and it was even already implemented once in #2283.

Having a yardmap (i.e. being a "true" building as opposed to merely an "immobile" like a nanotower) is exclusive with being mobile for similar reasons as mexes, i.e. the yardmap represents the capability to modify terrain in an expensive (pathing etc) and enduring way, while also coming with responsibilities (yardmap can only be rotated along cardinal directions and has to be aligned to the grid). In practice it looks like `CUnit` already contains a yardmap pointer and everything is handled at that level anyway  though.

Factories are a tough nut to crack. Things to ponder:
 * factory is really two capabilities: having a "produce things" queue, and having a rally order queue (i.e. one that the unit does not itself act upon). The former is independently useful without the latter (think starcraft evolution chamber or nuke silo), and so is the latter without the former (think starcraft nydus or bunker, for units exiting them).
 * there exists the capability of having a regular, non-rally (i.e. one that the unit acts upon itself) order queue. Non-factory units all have this capability. Being a builder, mobile, or armed sort of requires this capability (since build/move/attack commands have to go somewhere).
 * rally queue capabilities can be duplicate. Starcraft hatcheries have two rally queues (one for drones and another for everything else). The 2026 community balance patch for starcraft adds a third rally for queens.
 * the "produce things" queue capability can be duplicate. See starcraft reactor.
 * notice that for my examples I haven't gone beyond starcraft, and haven't even mentioned protoss (tfu!), so this is nothing out of ordinary.
 * Recoil units have two queue "slots". This means being a factory and being about anything else preclude each other since it would require 3+ queues. Adding a 3rd queue is not trivial because it requires some design work.

So I think:
 * folding `CBuilding` to `CUnit` sounds good, as long as having both a yardmap and mobility are rejected. 
 * it would be wise to keep `CFactory` as-is for now.
 * adding an `isFactory` tag that can be set to `false` for units that fulfil the current conditions to make them resolve to a `CBuilder` instead (while keeping the existing conditions as requirements for `true`) sounds good. With `CBuilding` folded into `CUnit` this should achieve what this ticket asks for, i.e. the possibility to combine builders with yardmaps that doesn't resolve to a factory.
 * turning `CExtractorBuilding` into a capability sounds ok, as long as it's rejected for mobiles (this means nanotower style immobiles would still become able to get it).
 * turning `CBuilder` into a capability sounds ok, although if the above are done that may well be lots of work with little benefit.

**Comment by burnhamrobertp** [comment]
> folding CBuilding to CUnit sounds good, as long as having both a yardmap and mobility are rejected.

Initially I had this implemented such that the unit def just failed to load if it didn't mean this criteria; but that seemed like it would effectively start erroring on (improperly defined) units like the ones [this PR](https://github.com/beyond-all-reason/Beyond-All-Reason/pull/9243) removed? So instead there's a logged deprecation warning on the first instance (as opposed to every instance which would just produce log spam).

**Comment by sprunk** [comment]
Sure, I meant rejecting physically having both capabilities. Resolving to either one is fine.

**Comment by burnhamrobertp** [comment]
With the first 3 PRs done, I was ruminating more on the CFactory (and FactoryCAI) complexities/challenges....every non-factory unit has a queue it acts upon itself, and being a builder, mobile or armed needs one because those orders have to go somewhere. A factory doesn't have one. FactoryCAI keeps the build queue in commandQue, the same member every other unit uses for its own orders, so there's nowhere for an order a factory would carry out itself to land. That's the same wall whether the order is repair, guard, reclaim or move, and it's also the ceiling on the class work: anything that's a factory plus something else needs a queue that doesn't exist.

CommandQueue.h:19 already has a QueueType of CommandQueueType, NewUnitQueueType and BuildQueueType, and FactoryCAI.cpp:59 sets it on both queues, but nothing ever reads it back - GetType() has no callers and creg just saves the value. So the first question is probably whether QueueType is meant to mean something or should just go, because a variable number of queues wants a key and right now that enum is either the obvious candidate or dead code someone should delete.

If it's worth keeping, would a variable count keyed on it be the shape? A unit gets a queue it acts upon itself when it needs one, a factory additionally gets a build queue and a rally queue, and a second rally queue stops being a special case.

It's been a long time since I wrote C++ with really finely-tuned memory use in mind, but I think a small vector of queues that starts empty could work without inflating anything? This particular detail is the one I'm probably least certain on in terms of viability? Regardless of the precise data structure, though, I expect there is _something_ that can be chosen that affords the necessary flexibility without expanding the sizing of unit data excessively.

The Lua surface is most of the work either way. GetUnitCommands, GetFactoryCommands, GetCommandQueue, GetFullBuildQueue, GetRealBuildQueue and the counts would want to become one accessor taking a queue id, with the current ones kept as shims. Routing is the other one - naming a target queue reads better than OPT_PUT_ORDER_INTO_THE_OPPOSITE_QUEUE once there can be more than two, though orders coming off the wire probably want the CommandAI to keep choosing by default. Fortunately, these functions do take care of the hardest parts of facilitating implementing this without introducing unavoidable breakages at the game boundary.

@sprunk would appreciate your thoughts as to the viability here. Obviously, its not strictly necessary but is an architectural improvement, I would argue, and opens up the opportunities for interesting possibilities in the future (a unit that's also a mobile factory, producing its own units as it slowly ambles about the battlefield, etc.)

**Comment by sprunk** [comment]
I'm not sure. There's many design questions with multiple answers with tradeoffs.

Consider the pre-game queue widget. It's conceptually a queue, but it's not actually bound to any unit (it larps as if it's for the commander, but it's emulated because the commander doesn't yet exist) and it's unsynced (it sends messages around but doesn't need to). Would it be good for queues not to be necessarily bound to a unit to be possible? To be unsynced?

Consider metacommands (insert and remove), and the network protocol. Where does the queueID go? How much of this handling should the engine relegate to games?

Consider the various weird commands. State commands (e.g. repeat toggle) are given directly to a unit and never actually end up in any queue, how does the API handle this? Factory build orders have some special rules (e.g. they don't require shift to be appended at the end, they cancel via rightclick), is this a property of the command or of the queue?

So, more concretely:
 * the QueueType enum may well turn out to be an obvious candidate for further use or dead code, but I think it is hard to tell yet at this point.
 * implementation details (vector vs whatever else) are going to depend on other choices above. Whether the three native queues live in the data structure or are kept as explicit members is an open question.
 * some of the questions above have answers that I like more than the others, but I don't yet have a full picture.

**Comment by burnhamrobertp** [comment]
I wouldn't think that one exception makes the case that queues have enough value existing as an abstraction completely decoupled from units. Admittedly, you might be mentioning the pre-game queue widget just as one example amongst dozens that I'm ignorant of?

For the metacommands and the protocol, I'd really need to do more research to provide you with a cognicent answer; I'm not familiar with the protocol side to speak to the implications of adding on a queueId or whatnot. You might be able to preserve existing backwards compatibility by having an implementation produce the same behavior as what you have right now (by assuming that any command that doesn't specify a queue targets the same queue that would've been targeted for that class of unit before the changes; but even with that stated I'm not sure what other exceptions there might be).

The state commands thing is an interesting point; thinking about it made me think of something I hadn't before that _could_ be done as part of a targeted refactor / rewrite of queues. Various states are set on a factory, but actually belong to the production queue in order to drive the initial values for the units produced by that queue. For example, consider a factory and its `Fire State` that has a turret on top. It is not possible (or at least, not engine-supported) right now to have such a unit set to fire at will but producing units with hold fire, because the data model just doesn't support it. I'm not sure that changing this wouldn't complicate the data model more than the value its adding, though? Take this with a grain of salt, perhaps its just an interesting thing to keep in mind, I don't have the perspective to definitely say that its valuable in and of itself.

> Factory build orders have some special rules (e.g. they don't require shift to be appended at the end, they cancel via rightclick), is this a property of the command or of the queue?

These seem like they have obvious answer to me, which suggests you're meaning to use them as an example to make a point that I'm missing? Or perhaps you're asking them because the answers, however obvious, are contentious for some reason I'm unaware of? Nevertheless, I would say that these "exceptions" are not exceptions at all, but rather normal baseline behaviors for this specific queue type and - more precisely - the product of specific interactions the player is making with the queue even if they think they're interacting with the unit.

All of the above in mind:

* Not only could QueueType be valuable, I think there's additional value hidden in it because queues of a certain type carry other certain properties (production queues can carry values for produced units' default states) AND implied behaviors when interacting with that particular type of queue.
* Is adding a 3rd queue solving the problem or just kicking the can down the road? Conversely, would it be overengineering to allow a unit to have `n` queues considering there -presumably- isn't anyone actively wanting to have a unit that can walk around, shoot, build buildings, produce things independently and simultaneously (i.e. units and anti-nuke ordinance, for example)?

**Comment by tomjn** [comment]
I come here with some historical context and notes that may put a spin on this

> One consequence of this is that we will need to decide what counts as factory. One option is a factory is a unit with buildpower, a buildlist, and can start construction within its own footprint. Thus, SupCom's Fatboy would count as a factory, since it builds units inside itself, but BAR's combat engineers wouldn't, since they build units beside themselves.

Why not both, expand and exterminate has units that build other units, as does nanoblobz, and one of the annoying things is that you can't build another unit in place then have it walk away and another unit starts construction, you had to place them in rows with repeat turned on, then move the units out the way so that the next batch would work.

I can foresee that a builder may want to build both structures and other units.

> Mexes are a subset of buildings because it is not well-defined what should happen when they move. 

This is why that was originally not implemented when it was first suggested, the cost of updated the metalmap. Cheaper than updating the terrain map and associated pathing, but not trivial eitherway.

---

I would also be mindful of what this all looks like in a unitdef, there's a lot of careful logic in gadgets and native AIs to figure out if a unit is a factory or not, and to account for nanotowers.

---

Technically a solution to one of the edge cases already exists:

https://springrts.com/phpbb/viewtopic.php?t=2604

Sadly the original thread of zwzsg's is lost to time, but you can see here in 2005 they were discussing this and the original experiment.

A static factory would build a unit that was larger than itself. That unit would then act as a transport and pick up the factory and act as its legs or wings and walk/fly somewhere else. The factory carried on doing factory things and the unit carried it on its back.


I'm happy to see consolidation of the internals, but I think these are the most realistic outcomes:

 - movable or deployable metal extractors
 - the ability to put a builder in factory mode
   - maybe using a new command that configures or places its "build plate"  so the engine knows where to position the unit and that the units in factory mode to pass commands on to the units it builds?

**Comment by tomjn** [comment]
closest to this I could find from the founder of the spring engine https://springrts.com/phpbb/viewtopic.php?p=26713#p26713

**Comment by sprunk** [comment]
> pre-game queue widget just as one example amongst dozens that I'm ignorant of?

* pre-game queue
* ferry routes (automated transport like in supcom1)
* orders for a whole team (think RA3 co-op campaign where you can order your partner if playing solo with a bot)
* persistent queue that survives if a fac dies and gets rebuilt
* unit queues for units you don't own and just control intermittently (think wc3 mercenary camp)
* formations like in supcom2

>> Factory build orders have some special rules (e.g. they don't require shift to be appended at the end, they cancel via rightclick), is this a property of the command or of the queue?
>
> These seem like they have obvious answer to me, which suggests you're meaning to use them as an example to make a point that I'm missing? Or perhaps you're asking them because the answers, however obvious, are contentious for some reason I'm unaware of? Nevertheless, I would say that these "exceptions" are not exceptions at all, but rather normal baseline behaviors for this specific queue type

I'm asking because the answers aren't obvious, there's arguments not to do things at the level of queue. Being stateful (as opposed to ending up in a queue) is defined in the command, and perhaps at this level of abstraction it has to be (because the decision to route a command to some queue necessarily cannot yet involve a queue). More generally, commands are more fundamental than queues and at least there's an existing interface to define their traits. The queue types are definitely useful but they could live in basecontent as a reference implementation, rather than be hardcoded at engine level.

> be mindful of what this all looks like in a unitdef, there's a lot of careful logic in gadgets and native AIs to figure out if a unit is a factory or not, and to account for nanotowers.

Assumptions in existing code sounds like a concern for games who would want use the new features. I think the engine only needs to make sure the relevant interfaces make sense and are reasonably backwards compatible where possible.

**Comment by burnhamrobertp** [comment]
> persistent queue that survives if a fac dies and gets rebuilt
> unit queues for units you don't own and just control intermittently (think wc3 mercenary camp)

Ferry routes did occur to me, and it was one I thought could fit nicely into the implementation I proposed above, at least conceptually. This is a transport's command queue with load and unload commands (with maybe a move in between?) and the command queue has the property/config `repeat == true`

> orders for a whole team (think RA3 co-op campaign where you can order your partner if playing solo with a bot)

I'm not 100% sure this is a queue, even if you extend RA3's AI coop such that it could support a series of specified actions? However, if we take this as the baseline and then increase the functionality several times over, I would contend that what this is better represented by is a queue on another item (the ally team) not a global queue object that exists in isolation.

> persistent queue that survives if a fac dies and gets rebuilt

Rebuilt? or resurrected? Rebuilt seems concerning unless the engine is meant to provide 1st class support for games that - for example - only allow 1 instance of each type of factory to exist? Resurrected is a different class of problem; the persistence of a unit after its death so that resurrection restores all manner of things (such as the previous command queue) automatically.

> unit queues for units you don't own and just control intermittently (think wc3 mercenary camp)

Maybe? Or if this is a practical desire, is this evidence of a new type of queue on the unit itself that has space for commands to carry ownership data?

> formations like in supcom2

Did supcom2 have formations? Forged alliance did, but they were automatic, I thought? In any case, a group of units' selected formation is a property of that group of selected units; not their orders, and not even the queue, I would contend? In BAR, its even more simplistic than that: the formation is a macroscopic property that is derivable only from the individual orders in each units' queue.

**Comment by tomjn** [comment]
> ferry routes (automated transport like in supcom1)

Area load and unload commands were added as a direct response to the ferry routes in supcom

**Comment by sprunk** [comment]
> Area load and unload commands were added as a direct response to the ferry routes in supcom

Yes, they don't offer as good UX as a dedicated queue though. The route is gone if the last transport gets shot down. There isn't an easy way to assign a transport to handle the route, or to reliably make sure all transports handling the route adjust to any changes to the route. There is no easy way to visualise the route without dragging in a bunch of caveats and processing.

> Did supcom2 have formations? Forged alliance did, but they were automatic, I thought? In any case, a group of units' selected formation is a property of that group of selected units; not their orders, and not even the queue, I would contend? In BAR, its even more simplistic than that: the formation is a macroscopic property that is derivable only from the individual orders in each units' queue.

I meant the ad hoc control group that forms when units congregate. I guess you could also just manage regular control groups this way. So rather than giving each unit the same commands and hope they stick together, units would just have a "follow the group" command and the wupget handling the group would take care to do things such as waiting for stragglers at nodes before resolving move commands (as is attempted by Recoil's "gather mode"), or arranging units into formation along the way.

> concerning unless the engine is meant to provide 1st class support for games that - for example - only allow 1 instance of each type of factory to exist?

The engine is meant to provide fundamental tools and "vocabulary" interfaces and let games use them to build things on top of that as they see fit. RA2 had global build queues and the production buildings were mostly just exit points. Campaign enemy AI could have a "master" queue that its factories follow. The "master" queue being virtual and surviving death of all subordinates is the specific trait I bring up because it's the one that can't be worked around by just copying queues around on actual units.

> I would contend that what [RA3 co-op queue] is better represented by is a queue on another item (the ally team) not a global queue object that exists in isolation.
>
> is [WC3 mercenary camp] evidence of a new type of queue on the unit itself that has space for commands to carry ownership data?

Possibly. We will see when we get there. I am mostly bringing up these examples in hopes that anybody picking up work here (not necessarily specifically you) would be aware of them and not block them off with some naive implementation that takes bad tradeoffs.
